For #24333 - Migrate CollectionViewHolder to Compose

Mugurell opened the pull request:
Fixes #24333

https://user-images.githubusercontent.com/11428869/160660421-c2a0978f-1179-416a-b835-13e69aa57d5e.mp4

Before and after layout bounds:

| Before | After |
| --- | --- |
| <img src="https://user-images.githubusercontent.com/11428869/164164597-d09d667f-a693-4a56-a204-52cd9ddcbba3.png" width="98%"> | <img src="https://user-images.githubusercontent.com/11428869/164167525-c8b4643d-c892-4f4b-ab9f-a1c86afc46c3.png"> |


### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

Mugurell commented on app/src/main/java/org/mozilla/fenix/compose/list/ExpandableListHeader.kt:
```diff
@@ -45,16 +45,21 @@ fun ExpandableListHeader(
     expandActionContentDescription: String? = null,
     collapseActionContentDescription: String? = null,
-    onClick: () -> Unit = {},
+    onClick: (() -> Unit)? = null,
```
These changes were needed to be able to handle clicks in the parent and in so show the expected indication.
Otherwise, the indication (and clicks) would look like

https://user-images.githubusercontent.com/11428869/160660920-d6c535b8-6db0-45fd-9a18-4081fc2bc400.mp4

MozillaNoah replied:
that's a weird interaction. thanks for the fix. this will also make it so any headers without click lambdas won't show the click ripple.

gabrielluong replied:
Extracted this PR out in https://github.com/mozilla-mobile/fenix/pull/24898

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt:
```diff
@@ -534,11 +526,12 @@ class HomeFragment : Fragment() {
-            * the View action on the [TabsTrayDialogFragment] snackbar.*/
-            scrollAndAnimateCollection(bundleArgs.getLong(FOCUS_ON_COLLECTION, -1))
         }
+        // else if (bundleArgs.getLong(FOCUS_ON_COLLECTION, -1) >= 0) {
```
Should this just be deleted?

Mugurell replied:
Yes, the functionality that this would call was dropped. Thank you!

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/home/collections/CollectionMenu.kt:
```diff
@@ -0,0 +1,142 @@
+
+    // DropdownMenu uses the medium shape from MaterialTheme.
+    // Override it's corner radius to be the same 8.dp as in mozac_browser_menu_corner_radius
+    MaterialTheme(shapes = MaterialTheme.shapes.copy(medium = RoundedCornerShape(8.dp))) {
```
Should we still be using FireFoxTheme here?

Mugurell replied:
`CollectionMenu` is expected to be in a composition with `FirefoxTheme` applied because it is hardcoded to read `FirefoxTheme.colors`.
`DropDownMenu` is expected to be in a composition with `MaterialTheme` applied because it is implemented as a `Card` which [is hardcoded to use `MaterialTheme.shapes.medium`](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:compose/material/material/src/commonMain/kotlin/androidx/compose/material/Card.kt;l=61;drc=4f2a25b8f8ee02053c57b524a1a6341e1edb9c06) as it's shape for which here we override the actual property.

While not pretty we will have to do something like this in multiple places in our app where we'd use android composables but with our own styling on top and so needing to override the defaults.

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/home/collections/CollectionMenu.kt:
```diff
@@ -0,0 +1,142 @@
+                        item.onClick()
+                    },
+                ) {
+                    Text(
```
does this `Text` need any text size attributes to be set here?

Mugurell replied:
Don't think so. This menu follows the other used for homescreen composables like the one [for RecentVisits](https://github.com/mozilla-mobile/fenix/blob/f924f744348006e2cb01fb172486fc7bfb1b67dc/app/src/main/java/org/mozilla/fenix/home/recentvisits/view/RecentlyVisited.kt#L328-L335), [for RecentBookmarks](https://github.com/mozilla-mobile/fenix/blob/f924f744348006e2cb01fb172486fc7bfb1b67dc/app/src/main/java/org/mozilla/fenix/home/recentbookmarks/view/RecentBookmarks.kt#L205-L212) or [for RecentTabs](https://github.com/mozilla-mobile/fenix/blob/f924f744348006e2cb01fb172486fc7bfb1b67dc/app/src/main/java/org/mozilla/fenix/home/recenttabs/view/RecentTabs.kt#L488-L495).

MozillaNoah replied:
gotcha gotcha. I just get a little anxious seeing key parameters not initialized 😅

Mugurell replied:
I know 😅. A reusable `Menu` / `MenuItem` might help with this, with them hardcoding a certain style to also ensure nothing changes in how the menu looks if some defaults change in Compose.

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/home/collections/CollectionMenu.kt:
```diff
@@ -0,0 +1,142 @@
+ * @param onDismissRequest Called when the user chooses a menu option or requests to dismiss the menu.
+ */
+@Composable
+fun CollectionMenu(
```
can we get a `@Preview` for this?

MozillaNoah replied:
I've had this issue occur usually because of the `components` object, in case that's what's causing it here. I may poke around later to also take a look.

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/home/collections/CollectionViewHolder.kt:
```diff
@@ -0,0 +1,194 @@
+
+    companion object {
+        val LAYOUT_ID = View.generateViewId()
+        const val maxTitleLength = 20
```
is this constant used anywhere?

Mugurell replied:
This was [inherited from the previous ViewHolder](https://github.com/mozilla-mobile/fenix/blob/f924f744348006e2cb01fb172486fc7bfb1b67dc/app/src/main/java/org/mozilla/fenix/home/sessioncontrol/viewholders/CollectionViewHolder.kt#L109) and seems to be used in [TabCollectionStorage](https://github.com/mozilla-mobile/fenix/blob/f924f744348006e2cb01fb172486fc7bfb1b67dc/app/src/main/java/org/mozilla/fenix/components/TabCollectionStorage.kt#L112) .

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/home/collections/CollectionViewHolder.kt:
```diff
@@ -0,0 +1,194 @@
+private val expandedCollectionShape = RoundedCornerShape(topStart = 8.dp, topEnd = 8.dp)
+
+@Composable
+private fun Collection(
```
this file feels pretty large with the Composable living here. could we possibly get it in its own file?

Mugurell replied:
Sure.

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/home/collections/CollectionViewHolder.kt:
```diff
@@ -0,0 +1,194 @@
+    collection: TabCollection,
+    expanded: Boolean,
+    menuItems: List<CollectionMenuItemData>,
+    interactor: CollectionInteractor,
```
similar to `CollectionMenu`, could you add a `@Preview` for this too?

Mugurell replied:
Will do.

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/home/collections/CollectionViewHolder.kt:
```diff
@@ -0,0 +1,194 @@
+                    end = 8.dp // (24.dp - 16.dp) hardcoded in ExpandableListHeader
+                ),
+                tint = Paint().apply {
+                    color = Color(collection.getIconColor(LocalContext.current))
```
should we maybe get a function in `FirefoxTheme` that handles this? that way we also can just do something like `tint = FirefoxTheme.colors.getCollectionColor()`

Mugurell replied:
The color seems to [depend on the particular collection for which it will be used](https://github.com/mozilla-mobile/fenix/blob/f924f744348006e2cb01fb172486fc7bfb1b67dc/app/src/main/java/org/mozilla/fenix/ext/TabCollection.kt#L19).
Case in which the current approach of it being an extension function to the collection seems a bit cleaner by having a smaller scope - if specs change about the collection color we wouldn't need to modify the app theme.

MozillaNoah replied:
gotcha gotcha. part of my concern was this function call not working within a preview, so fingers crossed

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/home/collections/CollectionViewHolder.kt:
```diff
@@ -0,0 +1,194 @@
+                            Icon(
+                                painter = painterResource(R.drawable.ic_share),
+                                contentDescription = stringResource(R.string.share_button_content_description),
+                                tint = FirefoxTheme.colors.textPrimary
```
I believe this should be `FirefoxTheme.colors.iconPrimary`

Mugurell replied:
Thank you!
`textPrimary` was [the color used in xml](https://github.com/mozilla-mobile/fenix/blob/f924f744348006e2cb01fb172486fc7bfb1b67dc/app/src/main/res/drawable/ic_share.xml#L11).
`FirefoxTheme.colors.iconPrimary` seems to have the same values as `FirefoxTheme.colors.textPrimary` and does fit better in the context, will change to that.

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/home/collections/CollectionViewHolder.kt:
```diff
@@ -0,0 +1,194 @@
+                                contentDescription = stringResource(
+                                    R.string.collection_menu_button_content_description
+                                ),
+                                tint = FirefoxTheme.colors.textPrimary
```
I believe this should be `FirefoxTheme.colors.iconPrimary`

Mugurell replied:
Thank you, will change to that.

MozillaNoah reviewed (commented):
first round of comments. I've tested locally and it seems to be working great so far!

gabrielluong commented on app/src/main/java/org/mozilla/fenix/home/HomeFragment.kt:
```diff
@@ -996,134 +989,6 @@ class HomeFragment : Fragment() {
-    /**
-     * Will highlight the border of the collection with the specified index.
-     * */
-    private fun animateCollection(indexOfCollection: Int) {
```
Did we discuss with Product or UX about whether or not they cared about keeping this interaction? Just want to know if we need to re-implement this or not.

Mugurell replied:
Removing this is a separate commit for which I've added in the description that a reassessment of the functionality is to happen in https://github.com/mozilla-mobile/fenix/issues/24389.

gabrielluong commented on app/src/main/java/org/mozilla/fenix/home/collections/CollectionViewHolder.kt:
```diff
@@ -0,0 +1,194 @@
+    collection: TabCollection,
+    expanded: Boolean,
+    menuItems: List<CollectionMenuItemData>,
+    interactor: CollectionInteractor,
```
We should avoid passing the `interactor` into any @Composable functions. Can you add their equivalent `onXXXX` callback function?

gabrielluong commented:
While there is no rush, I think we should avoid landing this until after the beta cut next Tuesday given that soft freeze is tomorrow.

Mugurell commented:
- Fixed the collections leakage.
Keeping the composable around to be recycled and binding new data in the viewholder to be observed for recomposition is just a cumbersome flow but it seemed to work in end.

- The shadow issue between the collection and the first containing tab is to be resolved in the migration of that containing tab item. The shadow is actually from this first tab, not from the collection.

<img src="https://user-images.githubusercontent.com/11428869/161098821-35ebbe4a-5dcd-46d9-83dc-bcffd2e375ff.png" width="33%" />

- Adding Oana as a reviewer for disabling the collections ui tests.

sv-ohorvath reviewed (approved):
Looks ok, we agreed it is better to comment out the code in this case because otherwise there would be too much to start over again.

gabrielluong commented:
Sorry for the delay, will try to get to this next week.

Mugurell commented:
Rebasing to rerun flaky ui test.

Mugurell commented:
Added new screenshots showing the layout bounds based on which when comparing side-by-side I saw that the chevrons `ExpandableListHeader` uses are the reversed of what's currently being used in xml and the reverse of that clicking the header will do. Added a new commit (second) to address this. @MozillaNoah .

gabrielluong commented:
Should be getting to this tomorrow.

gabrielluong commented on app/src/main/java/org/mozilla/fenix/compose/list/ExpandableListHeader.kt:
```diff
@@ -76,7 +76,7 @@ fun ExpandableListHeader(
                 Icon(
                     painter = painterResource(
-                        if (expanded) R.drawable.ic_chevron_down else R.drawable.ic_chevron_up
+                        if (expanded) R.drawable.ic_chevron_up else R.drawable.ic_chevron_down
```
![image](https://user-images.githubusercontent.com/1190888/164341755-72a65c7f-70d3-49d0-88ee-bb851a4532f6.png)

Not sure what to make of this right now since collections actually does the opposite of what the component library is proposing. Will check with @cwzilla tomorrow.

gabrielluong replied:
You're probably right though when I look at https://material.io/components/lists#specs. Again will check with @cwzilla tomorrow. I would generally recommend moving these types of fixes into a separate PR and having them land immediately since the changes are easy to verify and small in nature so very easy to review and just land.

gabrielluong commented on app/src/main/java/org/mozilla/fenix/home/collections/CollectionMenu.kt:
```diff
@@ -0,0 +1,198 @@
+@Composable
+fun getMenuItems(
+    collection: TabCollection,
+    onCollectionOpenTabsTapped: (TabCollection) -> Unit = {},
+    onRenameCollectionTapped: (TabCollection) -> Unit = {},
+    onCollectionAddTabTapped: (TabCollection) -> Unit = {},
+    onDeleteCollectionTapped: (TabCollection) -> Unit = {},
```
Since we always expect the consumer to provide a callback for these, I would remove the default parameter to enforce this.

gabrielluong commented on app/src/main/java/org/mozilla/fenix/home/collections/CollectionViewHolder.kt:
```diff
@@ -0,0 +1,103 @@
+ * @property isExpanded Whether the collection is expanded to show it's containing tabs or not.
+ */
+@Stable
+private data class CollectionInfo(
```
Can this just be `Collection`?

Mugurell replied:
Trying this rename I get
`Class 'CollectionInfo' will be shadowed by function 'Collection'`
as the Composable showing the collection already has this name.
There are quite a few `collection` related constructs. In this case `CollectionInfo` I think helps express a bit more about what it represents and help in knowing how and where to use it.

gabrielluong replied:
Ok. I was a bit effy about the `Info` suffix since we didn't any data classes that ended in that suffix in Fenix compare to `VisitInfo` in AC.

gabrielluong reviewed (commented):
This seems pretty good to me. Can you also rebase this since we landed some of your commits since they were good on their own? 

I am still wrestling with the timing of when to land this. Would landing all of your collections change after the beta cut sound good to you since the code freeze is this thursday?

Mugurell commented:
> This seems pretty good to me. Can you also rebase this since we landed some of your commits since they were good on their own? My one feedback would be to try to break standalone commits into their PR so we can just land that separately and more quickly instead of keeping the commit inside a big (900+ lines) PR. Also, one of your commits references the wrong issue number.

Rebased now and addressed your suggestions. Thank you!
Regarding breaking this in multiple patches I feel that actually means more work. 
This has sense as a whole and it should be easier and faster to work on, review and test as a whole.

> I am still wrestling with the timing of when to land this. Would landing all of your collections change after the beta cut sound good to you since the code freeze is this thursday?

I'm happy with what I can get. Agree that it would be better to wait a new train to have time to integrate and test all collection migrations and limit bigger issues.

Mugurell commented:
Rebased to fix conflicts.
@gabrielluong Was still left with two changes related to removing the xml border for xml collections which happened in https://github.com/mozilla-mobile/fenix/pull/25045.
Seems like that didn't contain the removal of `session_border.xml` and the dimens value it would use - `tab_border_width`. Not sure on whether there's a reason to keep them or not.

gabrielluong commented:
> Rebased to fix conflicts. @gabrielluong Was still left with two changes related to removing the xml border for xml collections which happened in #25045. Seems like that didn't contain the removal of `session_border.xml` and the dimens value it would use - `tab_border_width`. Not sure on whether there's a reason to keep them or not.

I only landed the piece that remove the collection animation in a separate PR since that was good to go. I couldn't remove those xml and dimen since those were still tied to the existing files. I am reviewing this today.

gabrielluong commented on app/src/main/java/org/mozilla/fenix/home/collections/CollectionViewHolder.kt:
```diff
@@ -0,0 +1,103 @@
+        val collectionInfo by remember { mutableStateOf(collectionData) }
+
+        collectionInfo.collection?.let { validCollection ->
+            val horizontalPadding =
+                composeView.resources.getDimensionPixelSize(R.dimen.home_item_horizontal_margin)
+            composeView.setPadding(horizontalPadding, 0, horizontalPadding, 0)
```
Can we move this out into the `init`? I think we already do this everywhere else.

gabrielluong commented on app/src/main/java/org/mozilla/fenix/home/collections/CollectionViewHolder.kt:
```diff
@@ -0,0 +1,103 @@
+
+    companion object {
+        val LAYOUT_ID = View.generateViewId()
+        const val maxTitleLength = 20
```
Can we move this out to where it's being called in `TabCollectionStorage` as a constant? Considering that the days for keeping `CollectionViewHolder` is probably limited we are probably prefer serve to refactor it now. I only bring this refactoring up because the review UI makes this file as appear as entirely new, even though I know we refactored it correctly.

gabrielluong commented on app/src/main/java/org/mozilla/fenix/home/collections/CollectionViewHolder.kt:
```diff
@@ -0,0 +1,103 @@
+    override fun Content() {
+        val collectionInfo by remember { mutableStateOf(collectionData) }
+
+        collectionInfo.collection?.let { validCollection ->
```
```suggestion
        collectionInfo.collection?.let { collection ->
```

gabrielluong commented:
Did we have also have a plan on addressing the line that appears at the bottom when it is expanded?

Mugurell commented on app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlAdapter.kt:
```diff
@@ -331,6 +335,11 @@ class SessionControlAdapter(
+            is CollectionViewHolder -> {
+                // Dispose the underlying composition immediately.
+                // This ViewHolder can be removed / re-added and we need it to show a fresh new composition.
+                holder.composeView.disposeComposition()
```
Good question.
Right now all composable viewholders will set a different than default strategy [here](https://github.com/mozilla-mobile/fenix/blob/58a6fafabc0477f6fa8f56adab8d37f849869e52/app/src/main/java/org/mozilla/fenix/compose/ComposeViewHolder.kt#L35-L37).
[DisposeOnViewTreeLifecycleDestroyed](https://developer.android.com/reference/kotlin/androidx/compose/ui/platform/ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed) rather than the default [DisposeOnDetachedFromWindow](https://developer.android.com/reference/kotlin/androidx/compose/ui/platform/ViewCompositionStrategy.DisposeOnDetachedFromWindow).

For them being inside a RecyclerView that may recycle holders the default would mean that when the holders are offscreen they would get recycled and the composition disposed 
to afterwards having to be recreate it when we'd need to show the same homescreen section again.
So setting `DisposeOnViewTreeLifecycleDestroyed` was a performance improvement. Create the holder, the composition and do not dispose it while the user is on the homescreen, only pay the penalty of creating these once.

This worked great for composable viewholders which would observe the store and update themselves.
In the case of the collections viewholders the update flow is a bit more convoluted. They will not observe the store but have the state injected.
In this case I had to go back to the default - `DisposeOnDetachedFromWindow` strategy to avoid them being still being shown on the screen after being deleted to prevent state leakage - the old holders showing stale data - what they previously displayed. This can be seen by commenting that line and deleting the collection / collection items and then undoing that action.
Keeping the default `DisposeOnDetachedFromWindow` in these cases means basically the same flow as for the xml version and ties nicely with the current way of passing data to these holders (highly convoluted, will probably need to be refactored later).
`CollectionViewHolder` could do something similar to [`ComposeViewHolder`](https://github.com/mozilla-mobile/fenix/blob/58a6fafabc0477f6fa8f56adab8d37f849869e52/app/src/main/java/org/mozilla/fenix/compose/ComposeViewHolder.kt#L35-L37) and override the strategy in it's init so we'd not have to do this `disposeComposition()` here but that would mean the holder knows where and how it is used, which it shouldn't, so I kept the default for it but managed the composition inside the adapter which is the actual component responsable for deciding what gets shown.

=> This approach is useful when state is passed from outside to avoid state leakage. 
We only need it because the compose strategy was overriden to be `DisposeOnViewTreeLifecycleDestroyed` and so holders will still remain on the screen until the screen gets destroyed.

Behavior without this:

https://user-images.githubusercontent.com/11428869/167076322-cd42dcd3-bae1-4396-9ffd-ce604e399fc7.mp4

Mugurell commented on app/src/main/java/org/mozilla/fenix/home/collections/Collection.kt:
```diff
@@ -0,0 +1,213 @@
+            .height(48.dp),
+        shape = if (isExpanded) expandedCollectionShape else collapsedCollectionShape,
+        backgroundColor = FirefoxTheme.colors.layer2,
+        elevation = 5.dp, // This needs to match the elevation of the TabInCollectionViewHolder for the shadow.
```
There is a matching comment in the new `TabInCollection` composable - https://github.com/mozilla-mobile/fenix/pull/24682/files#diff-fea09fe41897bb0e80f1fc96ae79d95b4fca055d5a548a5500fcb62a01ede580R118. With how problematic the shadows where I feel like it's an important detail. Will update the comment to rather point to `TabInCollection`. Can remov both if you think they aren't helping.

gabrielluong replied:
I would pick between removing it or referencing `home_item_elevation` since I would assume matching the card elevation between the two would be obvious, but maybe the best solution is just have a constant that is used by both. Looking ahead, were we thinking of joining the 2 ViewHolders in some way once we are done migrating both of them into Compose. For example, would it be possible to have this all under a single Card in a followup after we landed `TabInCollection`?

gabrielluong replied:
I also suspect that we should check with UX whether or not we should keep this at 5dp since all the other cards on Home is using 6dp, but this can be a followup.

Mugurell replied:
| For example, would it be possible to have this all under a single Card in a followup after we landed TabInCollection
In the future I'd hope so. Having all the collection details - header + items in a single Composable will probably help resolve many of the current issues with broken layouts / inconsistent state.
But this will require significant work to refactor not just the UI but the entire data flow so it will probably take some time to happen.
In the meantime maybe just removing that comment is best.

Mugurell commented on app/src/main/java/org/mozilla/fenix/home/collections/CollectionMenu.kt:
```diff
@@ -0,0 +1,210 @@
+        onDispose { onDismissRequest() }
+    }
+
+    // DropdownMenu uses the medium shape from MaterialTheme.
+    // Override it's corner radius to be the same 8.dp as in mozac_browser_menu_corner_radius
+    MaterialTheme(shapes = MaterialTheme.shapes.copy(medium = RoundedCornerShape(8.dp))) {
```
`DropdownMenu` [shows a `Popup` which shows a `DropdownMenuContent`](https://cs.android.com/androidx/platform/tools/dokka-devsite-plugin/+/master:testData/compose/source/androidx/compose/material/AndroidMenu.android.kt;l=95-106;drc=6fed3de7a56143de954d55e508a7449deb9af582).

`DropdownMenuContent` [shows a `Card`](https://cs.android.com/androidx/platform/tools/dokka-devsite-plugin/+/master:testData/compose/source/androidx/compose/material/Menu.kt;l=116;drc=6fed3de7a56143de954d55e508a7449deb9af582).

`Card` [uses by default `MaterialTheme.shapes.medium`](https://cs.android.com/androidx/platform/tools/dokka-devsite-plugin/+/master:testData/compose/source/androidx/compose/material/Card.kt;l=59;drc=6fed3de7a56143de954d55e508a7449deb9af582).

gabrielluong replied:
Thanks! I think this does warrant a followup issue to investigate the DropdownMenu styles vs BrowserMenu and ensuring we are aligning everywhere. Filed https://github.com/mozilla-mobile/fenix/issues/25149

Mugurell commented:
> Did we have also have a plan on addressing the line that appears at the bottom when it is expanded?

That seemed to actually be from the collection items.
Handled here https://github.com/mozilla-mobile/fenix/pull/24682/files#diff-fea09fe41897bb0e80f1fc96ae79d95b4fca055d5a548a5500fcb62a01ede580R193-R205.

Mugurell commented:
Thank for your help on this @gabrielluong!
Waiting on one small detail before merging - https://github.com/mozilla-mobile/fenix/pull/24501#discussion_r866523405.
